crates/shared/src/bad_token/token_owner_finder.rs

```diff
@@ -90,9 +106,82 @@ pub async fn init(
 
     if finders.contains(&TokenOwnerFindingStrategy::Liquidity) {
         if let Ok(finder) = BlockscoutTokenOwnerFinder::try_with_network(client.clone(), chain_id) {
-            result.push(Arc::new(finder));
+            proposers.push(Arc::new(finder));
         }
     }
 
-    Ok(result)
+    Ok(Arc::new(TokenOwnerFinder { web3, proposers }))
+}
+
+/// A `TokenOwnerFinding` implementation that queries a node with proposed owner candidates from an
+/// internal list of `TokenOwnerProposing` implementations.
+pub struct TokenOwnerFinder {
+    pub web3: Web3,
+    pub proposers: Vec<Arc<dyn TokenOwnerProposing>>,
+}
+
+impl TokenOwnerFinder {
+    /// Stream of addresses that might own the token.
+    fn candidate_owners(&self, token: H160) -> impl Stream<Item = H160> + '_ {
+        // Combine the results of all finders into a single stream.
+        let streams = self.proposers.iter().map(|finder| {
+            futures::stream::once(finder.find_candidate_owners(token))
+                .filter_map(|result| async {
+                    match result {
+                        Ok(inner) => Some(futures::stream::iter(inner)),
+                        Err(err) => {
+                            tracing::warn!(?err, "token owner proposing failed");
+                            None
+                        }
+                    }
+                })
+                .flatten()
+                .boxed()
+        });
+        futures::stream::select_all(streams)
+    }
+}
+
+#[async_trait::async_trait]
+impl TokenOwnerFinding for TokenOwnerFinder {
+    async fn find_owner(&self, token: H160, min_balance: U256) -> Result<Option<(H160, U256)>> {
+        let instance = ERC20::at(&self.web3, token);
+
+        // We use a stream with ready_chunks so that we can start with the addresses of fast
+        // TokenOwnerFinding implementations first without having to wait for slow ones.
+        let stream = self.candidate_owners(token).ready_chunks(MAX_BATCH_SIZE);
+        futures::pin_mut!(stream);
+
+        while let Some(chunk) = stream.next().await {
+            let mut batch = Web3CallBatch::new(self.web3.transport().clone());
+            let futures = chunk
+                .iter()
+                .map(|&address| {
+                    let balance = instance.balance_of(address).batch_call(&mut batch);
+                    async move {
+                        let balance = match balance.await {
+                            Ok(balance) => Some(balance),
+                            Err(err) if EthcontractErrorType::is_contract_err(&err) => None,
+                            Err(err) => return Err(err),
+                        };
+
+                        Ok((address, balance))
+                    }
+                })
+                .collect::<Vec<_>>();
+
+            batch.execute_all(MAX_BATCH_SIZE).await;
+            let balances = futures::future::try_join_all(futures).await?;
+
+            if let Some(largest_holder) = balances
```

Comment:
Right! Nice catch. I had made at some point the largest holder with `max_by_key` instead of `find`, but reverted the change and forgot to change the variable name.